graphix/opengraph.py

```diff
@@ -0,0 +1,162 @@
+"""Provides a class for open graphs."""
+
+from __future__ import annotations
+
+import math
+from typing import TYPE_CHECKING, NamedTuple
+
+import networkx as nx
+
+from graphix.generator import generate_from_graph
+
+if TYPE_CHECKING:
+    from collections.abc import Mapping
+
+    from graphix.pattern import Pattern
+    from graphix.pauli import Plane
+
+
+class Measurement(NamedTuple):
+    """An MBQC measurement.
+
+    :param angle: the angle of the measurement. Should be between [0, 2)
+    :param plane: the measurement plane
+    """
+
+    angle: float
+    plane: Plane
+
+    def isclose(self, other: Measurement, rel_tol=1e-09, abs_tol=0.0) -> bool:
+        """Compares if two measurements have the same plane and their angles
+        are close.
+
+        Example
+        -------
+        >>> from graphix.opengraph import Measurement
+        >>> from graphix.pauli import Plane
+        >>> Measurement(0.0, Plane.XY).isclose(Measurement(0.0, Plane.XY))
+        True
+        >>> Measurement(0.0, Plane.XY).isclose(Measurement(0.0, Plane.YZ))
+        False
+        >>> Measurement(0.1, Plane.XY).isclose(Measurement(0.0, Plane.XY))
+        False
+        """
+        return math.isclose(self.angle, other.angle, rel_tol=rel_tol, abs_tol=abs_tol) and self.plane == other.plane
+
+
+class OpenGraph:
+    """Open graph contains the graph, measurement, and input and output
+    nodes. This is the graph we wish to implement deterministically
+
+    :param inside: the underlying graph state
+    :param measurements: a dictionary whose key is the ID of a node and the
+        value is the measurement at that node
+    :param inputs: a set of IDs of the nodes that are inputs to the graph
+    :param outputs: a set of IDs of the nodes that are outputs of the graph
+
+    Example
+    -------
+    >>> import networkx as nx
+    >>> from graphix.opengraph import OpenGraph, Measurement
+    >>>
+    >>> inside_graph = nx.Graph([(0, 1), (1, 2), (2, 0)])
+    >>>
+    >>> measurements = {i: Measurement(0.5 * i, Plane.XY) for i in range(2)}
+    >>> inputs = [0]
+    >>> outputs = [2]
+    >>> og = OpenGraph(inside_graph, measurements, inputs, outputs)
+    """
+
+    __inside: nx.Graph
+    __inputs: list[int]
+    __outputs: list[int]
+    __meas: Mapping[int, Measurement]
```

Comment:
Now that inputs, outputs and measures are not stored in the graph, `OpenGraph` can now be declared as a `NamedTuple` with fields `inside`, `outputs` and `meas`, which would simplify the `__init__` implementation.

Reply:
Though if it in a NamedTuple, then we cannot write a custom `__init__` method, so we can't perform checks at object initialisation time. Currently there is only one check, namely that output nodes aren't measured.

We could get around this by making a classmethod to create new instances of `OpenGraph`, though we would need to keep in mind for all future code review that people should be using this method to create Open graphs and not directly with `OpenGraph(g, in, out, meas)`.

Something like
```python
@classmethod
def from(g: nx.Graph, in: list[int], out: list[int], meas: dict[int, Measurement]) -> OpenGraph:
    # Perform all the checks
    # ....
    return OpenGraph(g, meas, in, out)
```
As a side note, we also can't prefix attributes in a NamedTuple with double underscores so that change would have to be reverted.

Reply:
So in practice, only a few lines are saved in the `__init__` by using a NamedTuple, though we can remove the getter methods since it is now immutable. I think my main concern would be with having to ensure everyone uses the classmethod for initialisation from here one, which seems unnecessary.
What are your thoughts?

Reply:
The purpose of using double underscores is to protect these fields from being modified externally. However, since `NamedTuple` makes these fields immutable, we wouldn’t need to worry about protecting them, allowing us to eliminate the `__init__` method as well as the `inputs`, `outputs`, and `measurements` properties.

If the intention behind keeping `__init__` (and the other properties) is to enable validation, it might be more appropriate to use a `pydantic.BaseModel` with validators. That said, the current scope of validation seems limited. I’m unsure if it justifies keeping the existing setup. If we decide to maintain validation, it would be beneficial to include additional checks, such as ensuring that inputs, outputs, and measurements are actual graph nodes, and verifying that the graph is self-loop free.

Reply:
>  If we decide to maintain validation, it would be beneficial to include additional checks, such as ensuring that inputs, outputs, and measurements are actual graph nodes, and verifying that the graph is self-loop free.

@wlcsm could you add these? then we should be fine to resolve this conversation!

Reply:
Okay yep, I've made it a NamedTuple now. We now don't validate whether no output nodes were measured, but this can be added in later, perhaps with a more complete set of validations